core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -18,120 +18,234 @@
+pub trait StateFileIo {
+	type StateType;
+	type HashType;
 
-/// Writes the state (without the state diff) encrypted into the enclave storage
-/// Returns the hash of the saved state (independent of the diff!)
-pub(crate) fn write(state: StfState, shard: &ShardIdentifier) -> Result<H256> {
-	let state_path =
-		format!("{}/{}/{}", SHARDS_PATH, shard.encode().to_base58(), ENCRYPTED_STATE_FILE);
-	trace!("writing state to: {}", state_path);
+	/// Load a state file (returns error if it does not exist).
+	fn load(&self, shard_identifier: &ShardIdentifier, file_name: &str) -> Result<Self::StateType>;
 
-	// only save the state, the state diff is pruned
-	let cyphertext = encrypt(state.state.encode())?;
+	/// Create an empty (default initialized) state file.
+	fn create_initialized(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		file_name: &str,
+	) -> Result<Self::HashType>;
 
-	let state_hash = rsgx_sha256_slice(&cyphertext)?;
+	/// Write the state to file.
+	fn write(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		file_name: &str,
+		state: Self::StateType,
+	) -> Result<Self::HashType>;
 
-	debug!("new encrypted state with hash={:?} written to {}", state_hash, state_path);
+	/// Remove a state file.
+	fn remove(&self, shard_identifier: &ShardIdentifier, file_name: &str) -> Result<()>;
 
-	io_write(&cyphertext, &state_path)?;
-	Ok(state_hash.into())
-}
+	/// Checks if a given shard directory exists and contains state files.
+	fn shard_exists(&self, shard_identifier: &ShardIdentifier) -> bool;
 
-pub(crate) fn exists(shard: &ShardIdentifier) -> bool {
-	Path::new(&format!("{}/{}/{}", SHARDS_PATH, shard.encode().to_base58(), ENCRYPTED_STATE_FILE))
-		.exists()
+	/// Lists all shards.
+	fn list_shards(&self) -> Result<Vec<ShardIdentifier>>;
 }
 
-pub(crate) fn init_shard(shard: &ShardIdentifier) -> Result<()> {
-	let path = format!("{}/{}", SHARDS_PATH, shard.encode().to_base58());
-	fs::create_dir_all(path.clone())?;
-	let mut file = fs::File::create(format!("{}/{}", path, ENCRYPTED_STATE_FILE))?;
-	Ok(file.write_all(b"")?)
-}
+#[cfg(feature = "sgx")]
+pub mod sgx {
 
-pub(crate) fn read(path: &str) -> Result<Vec<u8>> {
-	let mut bytes = io_read(path)?;
+	use super::*;
+	use ita_stf::{State as StfState, StateType as StfStateType, Stf};
+	use itp_sgx_crypto::StateCrypto;
+	use itp_sgx_io::{read as io_read, write as io_write};
+	use itp_types::H256;
+	use log::*;
+	use sgx_tcrypto::rsgx_sha256_slice;
+	use std::string::ToString;
 
-	if bytes.is_empty() {
-		return Ok(bytes)
+	/// SGX state file I/O
+	pub struct SgxStateFileIo<StateKey> {
+		state_key: StateKey,
 	}
 
-	let state_hash = rsgx_sha256_slice(&bytes)?;
-	debug!(
-		"read encrypted state with hash {:?} from {}",
-		H256::from_slice(state_hash.as_ref()),
-		path
-	);
+	impl<StateKey> SgxStateFileIo<StateKey>
+	where
+		StateKey: StateCrypto<Error = itp_sgx_crypto::Error>,
+	{
+		pub fn new(state_key: StateKey) -> Self {
+			SgxStateFileIo { state_key }
+		}
```

Comment:
The symmetric state key (AES) is now provided upon construction of the `SgxStateFileIo` component and does not have to be read every time we do a state operation.

Reply:
Huh. I think we have a bit of a chaos here.

If we compare:
https://book.integritee.network/design.html

And our code:
https://github.com/integritee-network/worker/blob/master/enclave-runtime/src/tls_ra/seal_handler.rs#L60-L63

That doesn't match.

I'll create an issue for that. I think it might be related to the issue in https://github.com/integritee-network/worker/pull/602

Reply:
See https://github.com/integritee-network/worker/issues/703

Reply:
But anyway, you are correct. Lets leave as is for now, but it's something we need to adapt in the future if we want to have multiple shards per enclave I guess..

Reply:
Currently, I don't see the benefit of having multiple encryption keys. What scenario do you have in mind?

Reply:
Multisharding I guess?

Reply:
Created an issue: https://github.com/integritee-network/worker/issues/708